pkg/services/sqlstore/dashboard_version.go

```diff
@@ -67,49 +67,68 @@ func GetDashboardVersions(query *models.GetDashboardVersionsQuery) error {
 	return nil
 }
 
-const MAX_VERSIONS_TO_DELETE = 100
+const MAX_VERSIONS_TO_DELETE_PER_BATCH = 100
+const MAX_VERSION_DELETION_BATCHES = 50
 
 func DeleteExpiredVersions(cmd *models.DeleteExpiredVersionsCommand) error {
-	return inTransaction(func(sess *DBSession) error {
-		versionsToKeep := setting.DashboardVersionsToKeep
-		if versionsToKeep < 1 {
-			versionsToKeep = 1
-		}
+	return deleteExpiredVersions(cmd, MAX_VERSIONS_TO_DELETE_PER_BATCH, MAX_VERSION_DELETION_BATCHES)
+}
 
-		// Idea of this query is finding version IDs to delete based on formula:
-		// min_version_to_keep = min_version + (versions_count - versions_to_keep)
-		// where version stats is processed for each dashboard. This guarantees that we keep at least versions_to_keep
-		// versions, but in some cases (when versions are sparse) this number may be more.
-		versionIdsToDeleteQuery := `SELECT id
-			FROM dashboard_version, (
-				SELECT dashboard_id, count(version) as count, min(version) as min
-				FROM dashboard_version
-				GROUP BY dashboard_id
-			) AS vtd
-			WHERE dashboard_version.dashboard_id=vtd.dashboard_id
-			AND version < vtd.min + vtd.count - ?`
-
-		var versionIdsToDelete []interface{}
-		err := sess.SQL(versionIdsToDeleteQuery, versionsToKeep).Find(&versionIdsToDelete)
-		if err != nil {
-			return err
-		}
+func deleteExpiredVersions(cmd *models.DeleteExpiredVersionsCommand, perBatch int, maxBatches int) error {
+	versionsToKeep := setting.DashboardVersionsToKeep
+	if versionsToKeep < 1 {
+		versionsToKeep = 1
+	}
 
-		// Don't delete more than MAX_VERSIONS_TO_DELETE version per time
-		if len(versionIdsToDelete) > MAX_VERSIONS_TO_DELETE {
-			versionIdsToDelete = versionIdsToDelete[:MAX_VERSIONS_TO_DELETE]
-		}
+	for batch := 0; batch < maxBatches; batch++ {
+		deleted := int64(0)
+
+		batchErr := inTransaction(func(sess *DBSession) error {
+			// Idea of this query is finding version IDs to delete based on formula:
+			// min_version_to_keep = min_version + (versions_count - versions_to_keep)
+			// where version stats is processed for each dashboard. This guarantees that we keep at least versions_to_keep
+			// versions, but in some cases (when versions are sparse) this number may be more.
+			versionIdsToDeleteQuery := `SELECT id
```

Comment:
Could we move this outside the `inTransaction` call? We don't need the transactional integrity since the delete statement uses `In (?, ?, ?)` and will skip the already deleted rows.

Reply:
We certainly can, I just tried to make the changes as non-invasive as possible.